Sources/SmithyCBOR/Writer/Writer.swift

```diff
@@ -0,0 +1,224 @@
+                } else {
+                    try child.encode(encoder: encoder) // Recursively encode nested maps/arrays
+                }
+            }
+            encoder.encode(.indef_break)
+        } else {
+            // No value and no children: encode an empty map
+            encoder.encode(.indef_map_start)
+            encoder.encode(.indef_break)
+        }
+    }
+
+    public subscript(nodeInfo: NodeInfo) -> Writer {
+        if let child = children.first(where: { $0.nodeInfo == nodeInfo }) {
+            return child
+        } else {
+            let newChild = Writer(nodeInfo: nodeInfo, parent: self)
+            children.append(newChild)
+            return newChild
+        }
+    }
+
+    public func write(_ value: Bool?) throws {
+        guard let value else { return }
+        self.cborValue = .bool(value)
+    }
+
+    public func write(_ value: String?) throws {
+        guard let value else { return }
+        self.cborValue = .text(value)
+    }
+
+    public func write(_ value: Float?) throws {
+        guard let value else { return }
+        self.cborValue = .double(Double(value))
+    }
+
+    public func write(_ value: Double?) throws {
+        guard let value else { return }
+        self.cborValue = .double(value)
+    }
+
+    public func write(_ value: Int?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: Int8?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: Int16?) throws {
+        guard let value else { return }
+        self.cborValue = .int(Int64(value))
+    }
+
+    public func write(_ value: UInt8?) throws {
+        guard let value else { return }
+        self.cborValue = .uint(UInt64(value))
+    }
+
+    public func write(_ value: Data?) throws {
+        guard let value else { return }
+        self.cborValue = .bytes(value)
+    }
+
+    public func write(_ value: SmithyDocument?) throws {
+        // No operation.  Smithy document not supported in CBOR
+    }
+
+    public func writeTimestamp(_ value: Date?, format: TimestampFormat) throws {
+        guard let value else { return }
+        switch format {
+        case .epochSeconds:
+            self.cborValue = .date(value)
+        case .dateTime, .httpDate:
+            let string = TimestampFormatter(format: format).string(from: value)
+            self.cborValue = .text(string)
+        }
+    }
+
+    public func write<T>(_ value: T?) throws where T: RawRepresentable, T.RawValue == Int {
+        if let rawValue = value?.rawValue {
+            try write(rawValue)
+        }
+    }
+
+    public func write<T>(_ value: T?) throws where T: RawRepresentable, T.RawValue == String {
+        if let rawValue = value?.rawValue {
+            try write(rawValue)
+        }
+    }
+
+    public func writeMap<T>(
```

Comment:
map and array encoding is particularly complex due to the possibility of nested types. A map can be within an array and vice versa